### ISSUE TITLE
Love It.

### ISSUE THREAD

**Issue by LordDamocles** [DannyDeGaspari/Samsung-HVAC-buscontrol#3]
Hi Danny,

I've been searching for a solution to this since 2018. I have a Samsung MWR-WE10 wired control panel that I've been trying to setup remote access to (ideally into Hass.io). Until someone showed me your project my idea was just to have a servo push the power button on and off. I really want to implement what you've done here but this is way over my head. Is it at all possible to create a tutorial? I've already got a few RS485 on the way.

Thank you.
-Marcus

**Comment by DannyDeGaspari** [comment]
Hi Marcus,

Where exactly do you need help for ? I'm not familiar with Hass.io. Do you need help with the RS485 boards (did you order the same one as I used) ? Or do you need help with installing the scripts on a Raspberry Pi ?

Danny

**Comment by LordDamocles** [comment]
Hi Danny,

Thanks for getting back so soon. I should be able to get by with the Hass.io (Home Assistant) and scripts side of things. I'm a little lost on the RS485 and how to incorporate all this with the Pi. Here's the [RS485](https://www.ebay.com.au/itm/TTL-UART-to-485-RS485-RS422-Converter-Transceiver-Module-MAX485-RS232-OZ/202659409433?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2060353.m2749.l2648) I got.

**Comment by joeljody77** [comment]
Hey Danny.
Keen on this project too. Looking to get an arduino or esp8266 hooked up to bypass the cloud. Hoping to also to integrate into home assistant. Will have ago at this soon and see if I can make sense of it.
Cheers
Joel

**Comment by LordDamocles** [comment]
> 
> 
> Hi Marcus,
> 
> Where exactly do you need help for ? I'm not familiar with Hass.io. Do you need help with the RS485 boards (did you order the same one as I used) ? Or do you need help with installing the scripts on a Raspberry Pi ?
> 
> Danny

Hi Danny, could you please let me know if this is something you'd consider doing? My RS485 boards arrived today. Thanks.

**Comment by DannyDeGaspari** [comment]
Hi Marcus,

The board you bought is more difficult to control. It has DI, DE, RE, RO to connect to the PI's GPIO. The DI and RO can be connected to the UART RX and TX pins, the DE and RE can be connected together and act as an enable to control the flow for input or output. This control is something you have to do yourself and makes it less interesting. See this [link](https://circuitdigest.com/microcontroller-projects/rs485-serial-communication-between-arduino-and-raspberry-pi) for more info.

The board I use is [this](https://www.benl.ebay.be/itm/TTL-RS485-Adapter-485-UART-Seriell-3-3V-5-Volt-Level-Konverter-Modul-Arduino/252852473229?hash=item3adf2e8d8d:g:rBoAAOSwY~laD2WR) one. You do not need to care about this control signal and is easier to connect and use. Only the RX and TX pins to the UART of the PI and of course the power supply pins. 

The scripts in this repo do not control the DE and RE pins of the first board. So if you want to use it with your board, you will have to modify it.

By the way, nice to hear that all of you find this project useful.

Danny

**Comment by rabbit52068** [comment]
Hi Danny,

Thanks for making this really cool project, this is something that I've been searching for a while!
Would you mind to share how you have everything wired with some photos? Especially the part from the TTL convertor to the indoor unit. 

Thanks in advance!
Nelson

**Comment by DannyDeGaspari** [comment]
Hi Nelson,

Thanks. It is my plan to update this repo with more details and some pictures in the near future.

Danny

**Comment by vespix** [comment]
Hi Danny,

very nice job! I'm interesting in doing something similar.
Did you do any progress in understanding the command meaning?

Thank you!

**Comment by DannyDeGaspari** [comment]
Hi vespix,

Thank you. Meanwhile I updated the readme file with all the command tables I know so far. For me it is sufficient to have all control I need. However I would appreciate if anyone else can provide new info to complete the tables.
Good luck with your project.

Danny

**Comment by vespix** [comment]
Thank you for the update!
What do you mean for example with "bit 4-0 : room temperature - 9"?

Vespix

**Comment by DannyDeGaspari** [comment]
That field contains the room temperature in degrees celcius minus 9. See also the ac_status.py example program on how it is used.

**Comment by lanwin** [comment]
Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.

To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right? 

It would be cool to get a small ESP working with and simple interface like ESPHome.

**Comment by michelebergo** [comment]
> Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.
> 
> To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right?
> 
> It would be cool to get a small ESP working with and simple interface like ESPHome.

Hi Ianwin, I made it in esphome. if you want details you can contact me.

**Comment by lanwin** [comment]
> > Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.
> > To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right?
> > It would be cool to get a small ESP working with and simple interface like ESPHome.
> 
> Hi Ianwin, I made it in esphome. if you want details you can contact me.

@michelebergo yes I would really like to hear more. Whats the best way to contact you. Github did not have DM's?

**Comment by joeljody77** [comment]
I too would love an esphome solution for this hvac. Do away with the cloud
based approach.
Could i get some info as well ?

**Comment by LordDamocles** [comment]
@DannyDeGaspari 
Hey Danni, me again from 2020. Would also still love a guide on how to implement this.

**Comment by DannyDeGaspari** [comment]
Hey all,
Thank you for your interest. I too have a Home Assistant setup running, but the AC's are not controlled by it. So also for me it would be interesting in Michele's ESPHome solution. @michelebergo Can you share a link here ?

**Comment by michelebergo** [comment]
Hi all,
 if you want I can show you what I did over a skype call. I have not prepared a github project for it still.

> > > Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.
> > > To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right?
> > > It would be cool to get a small ESP working with and simple interface like ESPHome.
> > 
> > 
> > Hi Ianwin, I made it in esphome. if you want details you can contact me.
> 
> @michelebergo yes I would really like to hear more. Whats the best way to contact you. Github did not have DM's?

Hi all,
 if you want I can show you what I did over a skype call. I have not prepared a github project for it still.

**Comment by lanwin** [comment]
> Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.
> 
> > > > Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.
> > > > To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right?
> > > > It would be cool to get a small ESP working with and simple interface like ESPHome.
> > > 
> > > 
> > > Hi Ianwin, I made it in esphome. if you want details you can contact me.
> > 
> > 
> > @michelebergo yes I would really like to hear more. Whats the best way to contact you. Github did not have DM's?
> 
> Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.

@michelebergo I think it could hard to get everyone into a skype call. Would it also be possible to join in a private group chat? Maybe Skype or something different?

**Comment by michelebergo** [comment]
> > Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.
> > > > > Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.
> > > > > To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right?
> > > > > It would be cool to get a small ESP working with and simple interface like ESPHome.
> > > > 
> > > > 
> > > > Hi Ianwin, I made it in esphome. if you want details you can contact me.
> > > 
> > > 
> > > @michelebergo yes I would really like to hear more. Whats the best way to contact you. Github did not have DM's?
> > 
> > 
> > Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.
> 
> @michelebergo I think it could hard to get everyone into a skype call. Would it also be possible to join in a private group chat? Maybe Skype or something different?

@lanwin yes sure you can contact me on skype where we can create a chat group. I will send to you my username.

**Comment by michelebergo** [comment]
> > Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.
> > > > > Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.
> > > > > To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right?
> > > > > It would be cool to get a small ESP working with and simple interface like ESPHome.
> > > > 
> > > > 
> > > > Hi Ianwin, I made it in esphome. if you want details you can contact me.
> > > 
> > > 
> > > @michelebergo yes I would really like to hear more. Whats the best way to contact you. Github did not have DM's?
> > 
> > 
> > Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.
> 
> @michelebergo I think it could hard to get everyone into a skype call. Would it also be possible to join in a private group chat? Maybe Skype or something different?

my skype name --> michele.bergo

**Comment by LordDamocles** [comment]
@michelebergo 
Would it be possible at all to get a tutorial write up all? Just a normal step by step guide with the hardware you used?

**Comment by ksihota** [comment]
I have been following this thread in the hopes that it would allow me to control (or at least read power usage) of my 2 Samsung HVAC systems.
I am not clear how to do this and can't find anything useful from Samsung.
I have two units,
1 18K minisplit (1 output) with a single indoor head that is SmartThings enabled.
1 24K minisplit (3 outputs) using 2 indoor heads. One head is SmartThings enabled and the other is not. it uses a wired thermostat (MWR-WG00UN)
I can control 2 of the units through cloud control but would rather have these controlled locally. The wired unit has to be managed using the thermostat. 
I am currently converting all cloud systems to local control using HA. It would be nice to do this with my heat pumps as well. I have switched most firmware for devices to Tasmoda and am also using Zigbee for this I have a hub that access my main house power meter and provides power reading using an MQTT broker.
I have no problem playing around with cheap components but am a little concerned with making changes to a $12,000 HVAC system.

**Comment by lanwin** [comment]
@DannyDeGaspari here there are selly NASA and non NASA devices. As I understand it the use a different communication. For witch one is this project?

@michelebergo I send you an request via Skype.

**Comment by DannyDeGaspari** [comment]
> I have no problem playing around with cheap components but am a little concerned with making changes to a $12,000 HVAC system.

You are completely right. It is up to you to decide if you want to interfere into the Samsung electronics based on your own expertise. But if you use the method I used, it is rather robust. Though you still stay responsible for what you do.
I'm too into favour of local control on not depend on the cloud.

> @DannyDeGaspari here there are selly NASA and non NASA devices. As I understand it the use a different communication. For witch one is this project?

I don't know the difference and I also do not know which category my devices are into. I just reverse engineered the communication protocol without any other knowledge.

**Comment by lanwin** [comment]
I talked with my AC seller. All new devices are using the NASA protocol. The old devices without NASA will disappear soon. 

https://www.samsunghvac.com/Control-Non-NASA-Individual-Control/MWR-WE10

From the URL I would guess the MWR-WE10 is non NASA. So this protocol seems to be the non NASA version. 

@DannyDeGaspari I really hope you could help me reverse engineer the NASA based protocol when my devices are installed. So we could add them here.

**Comment by DannyDeGaspari** [comment]
Thanks for this information.

**Comment by WiktorBuczko** [comment]
> > > Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.
> > > > > > Hey Danny, thank you very much for that project. I am currently in the process of buying a 4 unit ac Samsung system and looking into how I can integrate this to HASS without using the Samsung Cloud.
> > > > > > To be clear? This is directly the binary bus protocol current NASA based units are talking. The MWR-WE10 is not required to make it work...right?
> > > > > > It would be cool to get a small ESP working with and simple interface like ESPHome.
> > > > > 
> > > > > 
> > > > > Hi Ianwin, I made it in esphome. if you want details you can contact me.
> > > > 
> > > > 
> > > > @michelebergo yes I would really like to hear more. Whats the best way to contact you. Github did not have DM's?
> > > 
> > > 
> > > Hi all, if you want I can show you what I did over a skype call. I have not prepared a github project for it still.
> > 
> > 
> > @michelebergo I think it could hard to get everyone into a skype call. Would it also be possible to join in a private group chat? Maybe Skype or something different?
> 
> @lanwin yes sure you can contact me on skype where we can create a chat group. I will send to you my username.

@michelebergo
Hi, I am also planing to integrate my Samsung  AC to Home Assistant using ESPHome, can I join your Skype chat group for more info on how to do it?

**Comment by lanwin** [comment]
I figured it out. NASA is a newer bus protocol. It uses similar start and stop bytes but the messages are more complex and transport more data. This is also why it uses 9600 baud instead of 2400. I also got it working directly with the F1,F2 bus protocol. 

I created an ESPHome component with support for NASA protocol witch works with F1,F2. I also made sure the Non NASA protocol can simply be integrated into this repo since this would allow us to use this component for both protocols. 

Currently it can show the room temperature, change the target temperature, set the AC mode and works with multiple units.

https://github.com/lanwin/esphome_samsung_ac

**Comment by lanwin** [comment]
Btw. Samsungs S-NET pro 2 app for NASA is written in dotnet an can simply be decompiled and reverse engineered. I did not tried it, but I bet that is also true for the Non NASA version of that app.
